host/host.go

```diff
@@ -27,46 +32,73 @@ type Host struct {
 
 // New creates a new Host.
 func New(log zerolog.Logger, address string, port uint, options ...func(*Config)) (*Host, error) {
-
 	cfg := defaultConfig
 	for _, option := range options {
 		option(&cfg)
 	}
 
 	hostAddress := fmt.Sprintf("/ip4/%v/tcp/%v", address, port)
-	addresses := []string{
-		hostAddress,
-	}
-
-	if cfg.Websocket {
-
-		// If the TCP and websocket port are explicitly chosen and set to the same value, one of the two listens will silently fail.
-		if port == cfg.WebsocketPort && cfg.WebsocketPort != 0 {
-			return nil, fmt.Errorf("TCP and websocket ports cannot be the same (TCP: %v, Websocket: %v)", port, cfg.WebsocketPort)
-		}
+	addresses := []string{hostAddress}
 
-		wsAddr := fmt.Sprintf("/ip4/%v/tcp/%v/ws", address, cfg.WebsocketPort)
-		addresses = append(addresses, wsAddr)
-	}
+	// define a subset of the default transports, so that we can offer a x509 certificate for the websocket transport
+	DefaultTransports := libp2p.ChainOptions(
+		libp2p.Transport(tcp.NewTCPTransport),
+		libp2p.Transport(quic.NewTransport),
+		libp2p.Transport(webtransport.New),
+	)
 
 	opts := []libp2p.Option{
-		libp2p.ListenAddrStrings(addresses...),
-		libp2p.DefaultTransports,
+		DefaultTransports,
 		libp2p.DefaultMuxers,
 		libp2p.DefaultSecurity,
 		libp2p.NATPortMap(),
 	}
 
 	// Read private key, if provided.
+	var key crypto.PrivKey
+	var err error
+
 	if cfg.PrivateKey != "" {
-		key, err := readPrivateKey(cfg.PrivateKey)
+		key, err = readPrivateKey(cfg.PrivateKey)
 		if err != nil {
 			return nil, fmt.Errorf("could not read private key: %w", err)
 		}
 
 		opts = append(opts, libp2p.Identity(key))
 	}
 
+	var tlsConfig *tls.Config
+	if cfg.Websocket {
+
+		// If the TCP and websocket port are explicitly chosen and set to the same value, one of the two listens will silently fail.
+		if port == cfg.WebsocketPort && cfg.WebsocketPort != 0 {
+			return nil, fmt.Errorf("TCP and websocket ports cannot be the same (TCP: %v, Websocket: %v)", port, cfg.WebsocketPort)
+		}
+
+		// Convert libp2p private key to crypto.PrivateKey
+		cryptoPrivKey, err := convertLibp2pPrivKeyToCryptoPrivKey(key)
```

Comment:
This here line relies on us having a key. However the key is optional - we could generate a libp2p identity on the fly.

If we start the node with `node --websocket true` (without `--private-key`) - this will blow up.

What I'm thinking:

1. We change how private key config works for the host. Instead of this:

```go
// WithPrivateKey specifies the private key for the Host.
func WithPrivateKey(filepath string) func(*Config) 
```

we do:

```go
// WithPrivateKey specifies the private key for the Host.
func WithPrivateKey(key crypto.PrivKey) func(*Config)
```

2. We make the key a hard requirement - `host.New` fails if it does not have a key.
3. In the `main()` function of the node, we take care of a) reading the key from disk - if the private key is specified, or b) we generate an in-memory key that we will use for the current invocation. In either case we pass the key to the `host`. Then we can rely on having the key.


What do you think?